app/database/operator/handlers/base_handler.ts

```diff
@@ -0,0 +1,457 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {MM_TABLES} from '@constants/database';
+import DatabaseManager from '@database/manager';
+import {
+    isRecordAppEqualToRaw,
+    isRecordCustomEmojiEqualToRaw,
+    isRecordGlobalEqualToRaw,
+    isRecordRoleEqualToRaw,
+    isRecordServerEqualToRaw,
+    isRecordSystemEqualToRaw,
+    isRecordTermsOfServiceEqualToRaw,
+} from '@database/operator/comparators';
+import {
+    prepareAppRecord,
+    prepareCustomEmojiRecord,
+    prepareGlobalRecord,
+    prepareRoleRecord,
+    prepareServersRecord,
+    prepareSystemRecord,
+    prepareTermsOfServiceRecord,
+} from '@database/operator/prepareRecords/general';
+import {Database, Q} from '@nozbe/watermelondb';
+import Model from '@nozbe/watermelondb/Model';
+
+import {
+    BatchOperationsArgs,
+    DatabaseInstance,
+    HandleEntityRecordsArgs,
+    HandleIsolatedEntityArgs,
+    PrepareForDatabaseArgs,
+    PrepareRecordsArgs,
+    ProcessInputsArgs,
+    RawValue,
+    RecordPair,
+} from '@typings/database/database';
+import {IsolatedEntities, OperationType} from '@typings/database/enums';
+
+import DataOperatorException from '../../exceptions/data_operator_exception';
+import DatabaseConnectionException from '../../exceptions/database_connection_exception';
+import {
+    getRangeOfValues,
+    getRawRecordPairs,
+    getUniqueRawsBy,
+    hasSimilarUpdateAt,
+    retrieveRecords,
+} from '../utils/general';
+
+export interface BaseHandlerMix {
+  activeDatabase: Database;
+  getActiveDatabase: () => DatabaseInstance;
+  setActiveDatabase: (database: Database) => void;
+  handleIsolatedEntity: ({tableName, values, prepareRecordsOnly}: HandleIsolatedEntityArgs) => boolean | Model[];
+  handleEntityRecords: ({findMatchingRecordBy, fieldName, operator, rawValues, tableName, prepareRecordsOnly}: HandleEntityRecordsArgs) => Promise<null | Model[]>;
+  processInputs: ({rawValues, tableName, findMatchingRecordBy, fieldName}: ProcessInputsArgs) => Promise<{ createRaws: RecordPair[]; updateRaws: RecordPair[] }>;
+  batchOperations: ({database, models}: BatchOperationsArgs) => Promise<void>;
+  prepareRecords: ({database, tableName, createRaws, updateRaws, recordOperator}: PrepareRecordsArgs) => Promise<Model[]>;
+  executeInDatabase: ({createRaws, recordOperator, tableName, updateRaws}: PrepareForDatabaseArgs) => Promise<void>;
+  getDatabase: (tableName: string) => Database;
+  getDefaultDatabase: () => Promise<Database>;
+  getServerDatabase: () => Promise<Database>;
+}
+
+class BaseHandler {
+  /**
+   * activeDatabase : In a multi-server configuration, this connection will be used by WebSockets and other parties to update databases other than the active one.
+   * @type {DatabaseInstance}
+   */
+  activeDatabase: DatabaseInstance;
+
+  constructor(serverDatabase?: Database) {
+      this.activeDatabase = serverDatabase;
```

Comment:
don't you need to set the current active database here if `serverDatabase` is undefined ? I know you are handling that in `getServerDatabase` but doing that can cause issues if `getActiveDatabase` returns **undefined**

Reply:
The DataOperator gets the currently active database from the Database manager.  Now we have the wrapper to which we pass in an active and valid connection.  The wrapper would create a valid connection or throw an error if it can't ).  Hence, only valid connections are passed to the Operator.

`this.activeServerDatabase` will be defined only if the Operator object has been instantiated by the wrapper.